envoy/http/codec.h

```diff
@@ -227,6 +194,56 @@ class RequestDecoder : public virtual StreamDecoder {
    * @return StreamInfo::StreamInfo& the stream_info for this stream.
    */
   virtual StreamInfo::StreamInfo& streamInfo() PURE;
+
+  /**
+   * @return Shared pointer to the stream_info for this stream.
+   */
+  virtual std::shared_ptr<StreamInfo::StreamInfo> streamInfoSharedPtr() PURE;
```

Comment:
I'm concerned about folks adding things to stream info that aren't also shared pointers, and will result in issues if the info outlives the stream.
Can you think of a way to test this?

Reply:
What do you think about a test similar to StreamInfoImplTest::SetFrom where we:

1) assert that the size of StreamInfo hasn't changed (if it has, print warning that new fields need to be shared)
2) set all the fields of a StreamInfo, have the original variables all go out of scope, assert that each of those fields still point to something.

Reply:
After a lot of tinkering with this, @danzh2010 and I decided not to make this a shared ptr in the end, and instead copy the stream info into the deferred-logging object at the time of stream destroy. This involves a new `setFrom` method similar to `setFromForRecreateStream`, and an expanded test in case StreamInfo gets new fields.